e2e suite: cross-file state pollution via shared Go server causes query-cache.test.ts flakes

## Problem

The e2e suite spins up a single Go server in `globalSetup` and shares it across every vitest test file. Because `vitest.config.ts` uses `pool: 'forks'` (default concurrency), multiple test files run in parallel against the same server — so any `TriggerRefresh` fired by one file pushes to active subscriptions in another file, corrupting assertions that depend on notification counts.

Surfaced on master at commit 58e65f9 (post-merge CI for PR #179):

```
❯ tests/query-cache.test.ts > Query Subscription Cache >
    updateCachedSnapshot notifies all subscribers with fresh data
AssertionError: expected 3 to be 2 // Object.is equality
- Expected: 2
+ Received: 3
 ❯ tests/query-cache.test.ts:354:30
```

The exact same merge SHA passed the e2e job on the PR's own CI run (24331319043) and failed on the post-merge master run (24331397241). A manual rerun of the failed job also passed. Classic flake signature.

## Root cause

`e2e/tests/setup.ts` spawns `go run ./server` in `globalSetup`, writes the address to `.test-server-addr`, and every test file reads that address via `helpers.ts`'s `getServerAddr()`. There is exactly one server for the whole suite.

Four test files call `createUser` in the happy path: `ws.test.ts`, `sse.test.ts`, `subscription.test.ts`, `query-cache.test.ts`. The vanilla handler at `example/vanilla/api/handlers.go:85` fires `aprot.TriggerRefresh(ctx, "users")` on every create. That trigger re-executes every active `ListUsers` subscription registered via `aprot.RegisterRefreshTrigger(ctx, "users")` at `example/vanilla/api/handlers.go:134`.

`query-cache.test.ts:315` ("updateCachedSnapshot notifies all subscribers with fresh data") subscribes to `listUsers:[]`, captures `notifyCountABefore`, fetches fresh data out-of-band, then calls `updateCachedSnapshot` and expects exactly one additional notification. When another test file calls `createUser` inside that window, the server pushes a refresh to the already-active cache subscription and `cbA` fires an extra time — count becomes `notifyCountABefore + 2` instead of `+1`, and the assertion fails with `expected 3 to be 2`.

The race window is short (tens of milliseconds, depending on scheduler), which is why it reproduces ~1 time in N on GitHub runners.

## Reproduction

Hard to reproduce deterministically because it depends on inter-process scheduling, but adding a small sleep between the subscribe and the assertion in `query-cache.test.ts:340-354` widens the window enough to trip it on most runs. The simpler evidence:

1. Same commit SHA yields different e2e outcomes across runs.
2. `grep -l "createUser" e2e/tests/*.ts` returns four files, all running in parallel against one server.
3. `example/vanilla/api/handlers.go:85` fires a global `TriggerRefresh("users")` on every `CreateUser`.
4. `query-cache.test.ts` is the only file whose assertions are sensitive to exact notification counts on `listUsers:[]`.

## Proposed fix

Give each test file (i.e. each vitest worker under `pool: 'forks'`) its own Go server process. Since there's no shared state between workers, there can be no cross-file refresh-trigger pollution.

Two-step setup to avoid paying `go run` compile time N times:

1. **`globalSetup`** — build the server binary once (`go build -o e2e/server/server ./server`). Runs before any test file.
2. **`setupFiles`** — one per-worker file that spawns the pre-built binary, reads its listen address from stdout, stashes it in a worker-local `globalThis.__APROT_E2E_SERVER_ADDR__`, and tears the process down in `afterAll`.
3. **`helpers.ts`** — `getServerAddr()` reads from the worker-local global instead of the filesystem. No `.test-server-addr` file; no singleton.

Trade-offs:

- **Pros:** true isolation, no `.test-server-addr` race conditions, `TriggerRefresh` calls stay inside the worker that fired them, each test file gets a clean user store.
- **Cons:** one extra Go process per test file during the run (~7 right now). Runtime cost is small because the binary is pre-built and each server takes <100ms to start. Memory: ~10-20 MB per process.

## Alternatives considered

1. **`singleFork: true`** — serializes test files to a single worker. One-line config change, keeps the shared server. Eliminates cross-file pollution because only one file runs at a time, but doubles wall-clock test time and hides the real isolation bug.
2. **Per-test unique trigger keys** — rewrite the test handlers to use a per-test-file trigger key so `TriggerRefresh` from one file can't reach another. Intrusive and fragile — every future test that touches a shared handler has to remember to do this.
3. **Retry the flaky test** — `test.retry(2)` on the specific assertion. Hides the symptom, lets the race linger.

Per-file isolation is the only fix that preserves parallelism, matches how the rest of the Go ecosystem isolates e2e state (one server per test binary), and doesn't require touching handler code.

## Affected versions

Present since `pool: 'forks'` + the shared-server `globalSetup` setup were introduced in #159 (the query cache PR). Not caused by, but surfaced by, PRs #177 and #179 because those were the first merges where the flake window lined up with post-merge CI.

--- a/.gitignore
+++ b/.gitignore
@@ -34,4 +34,6 @@ dist/
 example/vanilla/client/static/app.js
 example/react/client/dist/
 e2e/api/
-e2e/.test-server-addr
+
+# e2e server binary built by tests/setup.ts globalSetup (issue #180)
+e2e/server/server

--- a/e2e/tests/helpers.ts
+++ b/e2e/tests/helpers.ts
@@ -1,17 +1,22 @@
-import { readFileSync } from 'fs';
-import { dirname, join } from 'path';
-import { fileURLToPath } from 'url';
 import { WebSocket } from 'ws';
 import { EventSource } from 'eventsource';
-
-const __dirname = dirname(fileURLToPath(import.meta.url));
 
 // Polyfill browser APIs for Node.js test environment
 (globalThis as unknown as Record<string, unknown>).WebSocket = WebSocket;
 (globalThis as unknown as Record<string, unknown>).EventSource = EventSource;
 
+// Per-worker server address, populated by setup-per-file.ts's beforeAll
+// hook (issue #180). Reading from a worker-local global keeps every test
+// file pointed at its own server, so TriggerRefresh calls in one file
+// never leak into active subscriptions in another file.
 export function getServerAddr(): string {
-    return readFileSync(join(__dirname, '..', '.test-server-addr'), 'utf-8').trim();
+    const addr = globalThis.__APROT_E2E_SERVER_ADDR__;
+    if (!addr) {
+        throw new Error(
+            'Test server address not set — ensure ./tests/setup-per-file.ts is listed in vitest.setupFiles',
+        );
+    }
+    return addr;
 }
 
 export function wsUrl(): string {

--- a/e2e/tests/setup-per-file.ts
+++ b/e2e/tests/setup-per-file.ts
@@ -1,0 +1,81 @@
+import { type ChildProcess, spawn, execFileSync } from 'child_process';
+import { dirname, join } from 'path';
+import { fileURLToPath } from 'url';
+import { beforeAll, afterAll } from 'vitest';
+
+// Per-worker server lifecycle. Vitest runs this file once per test file
+// under pool: 'forks', so the beforeAll/afterAll hooks below apply to the
+// whole test file and each file gets its own Go server process — no
+// cross-file TriggerRefresh pollution (issue #180).
+//
+// The server binary itself is built once in setup.ts (globalSetup) so each
+// worker pays process-spawn cost only, not go-build cost.
+
+const __dirname = dirname(fileURLToPath(import.meta.url));
+const binaryPath = join(
+    __dirname,
+    '..',
+    'server',
+    process.platform === 'win32' ? 'server.exe' : 'server',
+);
+
+declare global {
+    // Worker-local because each vitest fork has its own globalThis. Reading
+    // from helpers.ts instead of the filesystem keeps the address scoped to
+    // the current test file's server.
+    // eslint-disable-next-line no-var
+    var __APROT_E2E_SERVER_ADDR__: string | undefined;
+}
+
+let serverProcess: ChildProcess | undefined;
+
+beforeAll(async () => {
+    serverProcess = spawn(binaryPath, [], {
+        cwd: join(__dirname, '..'),
+        stdio: ['ignore', 'pipe', 'pipe'],
+        // Detached on Unix so we can kill the whole process group on
+        // teardown; the pre-built binary is a single process today, but
+        // staying consistent with the prior setup is safer against future
+        // wrapper scripts.
+        detached: process.platform !== 'win32',
+    });
+
+    const addr = await new Promise<string>((resolve, reject) => {
+        let data = '';
+        serverProcess!.stdout!.on('data', (chunk: Buffer) => {
+            data += chunk.toString();
+            const newline = data.indexOf('\n');
+            if (newline !== -1) {
+                resolve(data.substring(0, newline).trim());
+            }
+        });
+        serverProcess!.stderr!.on('data', (chunk: Buffer) => {
+            process.stderr.write(chunk);
+        });
+        serverProcess!.on('error', reject);
+        setTimeout(() => reject(new Error('Server startup timeout')), 30000);
+    });
+
+    globalThis.__APROT_E2E_SERVER_ADDR__ = addr;
+}, 30000);
+
+afterAll(() => {
+    if (serverProcess?.pid) {
+        if (process.platform === 'win32') {
+            try {
+                execFileSync('taskkill', ['/f', '/t', '/pid', String(serverProcess.pid)], {
+                    stdio: 'ignore',
+                });
+            } catch {
+                // Process may already be gone
+            }
+        } else {
+            try {
+                process.kill(-serverProcess.pid, 'SIGKILL');
+            } catch {
+                serverProcess.kill('SIGKILL');
+            }
+        }
+    }
+    globalThis.__APROT_E2E_SERVER_ADDR__ = undefined;
+});

--- a/e2e/tests/setup.ts
+++ b/e2e/tests/setup.ts
@@ -1,59 +1,39 @@
-import { type ChildProcess, spawn, execSync } from 'child_process';
-import { writeFileSync, unlinkSync } from 'fs';
+import { execFileSync } from 'child_process';
 import { dirname, join } from 'path';
 import { fileURLToPath } from 'url';
+import { existsSync, unlinkSync } from 'fs';
 
 const __dirname = dirname(fileURLToPath(import.meta.url));
-const ADDR_FILE = join(__dirname, '..', '.test-server-addr');
 
-let serverProcess: ChildProcess;
+// Pre-build the Go e2e server once before any test file runs. Each test
+// file then spawns its own copy of this binary via setup-per-file.ts so
+// every vitest worker gets an isolated server — see issue #180 for the
+// cross-file TriggerRefresh pollution that broke query-cache.test.ts when
+// all workers shared a single server via globalSetup.
+const binaryPath = join(
+    __dirname,
+    '..',
+    'server',
+    process.platform === 'win32' ? 'server.exe' : 'server',
+);
 
 export async function setup() {
-    serverProcess = spawn('go', ['run', './server'], {
-        cwd: join(__dirname, '..'),
-        stdio: ['ignore', 'pipe', 'pipe'],
-    });
-
-    const addr = await new Promise<string>((resolve, reject) => {
-        let data = '';
-        serverProcess.stdout!.on('data', (chunk: Buffer) => {
-            data += chunk.toString();
-            const newline = data.indexOf('\n');
-            if (newline !== -1) {
-                resolve(data.substring(0, newline).trim());
-            }
-        });
-        serverProcess.stderr!.on('data', (chunk: Buffer) => {
-            process.stderr.write(chunk);
-        });
-        serverProcess.on('error', reject);
-        setTimeout(() => reject(new Error('Server startup timeout')), 60000);
-    });
-
-    writeFileSync(ADDR_FILE, addr);
+    execFileSync(
+        'go',
+        ['build', '-o', binaryPath, './server'],
+        {
+            cwd: join(__dirname, '..'),
+            stdio: 'inherit',
+        },
+    );
 }
 
 export async function teardown() {
-    if (serverProcess?.pid) {
-        // go run spawns a child process; kill the entire tree
-        if (process.platform === 'win32') {
-            try {
-                execSync(`taskkill /f /t /pid ${serverProcess.pid}`, { stdio: 'ignore' });
-            } catch {
-                // Process may already be gone
-            }
-        } else {
-            // Negative PID kills the process group on Unix
-            try {
-                process.kill(-serverProcess.pid, 'SIGKILL');
-            } catch {
-                serverProcess.kill('SIGKILL');
-            }
+    if (existsSync(binaryPath)) {
+        try {
+            unlinkSync(binaryPath);
+        } catch {
+            // Ignore if already cleaned up
         }
-    }
-    try {
-        unlinkSync(ADDR_FILE);
-    } catch {
-        // Ignore if already cleaned up
     }
 }

--- a/e2e/vitest.config.ts
+++ b/e2e/vitest.config.ts
@@ -2,10 +2,16 @@ import { defineConfig } from 'vitest/config';
 
 export default defineConfig({
     test: {
+        // globalSetup builds the Go server binary once before any test file.
+        // setupFiles then runs per-worker (once per test file under
+        // pool: 'forks') and spawns an isolated server process for that
+        // file — issue #180 eliminates cross-file TriggerRefresh pollution
+        // that caused flakes in query-cache.test.ts.
         globalSetup: './tests/setup.ts',
+        setupFiles: ['./tests/setup-per-file.ts'],
         testTimeout: 30000,
-        hookTimeout: 15000,
-        teardownTimeout: 1000,
+        hookTimeout: 30000,
+        teardownTimeout: 5000,
         pool: 'forks',
     },
 });

